Make extension operations lawful

harendra-kumar opened the pull request:
See the changelog for more details.

harendra-kumar commented:
TODO: allow trailing dots in extensions.

harendra-kumar commented:
All changes are done, this is ready for review. This change makes all extension operations fully principled which is a significant improvement, given how messy they can be in general. However, this change may break some existing code using extension operations. There is no change in API, only change is what is treated as a valid extension. The changelog mentions all breakages and how to handle it.  The documentation fully describes accurate behavior and several hundred test cases are added to test as many scenarios as I could imagine.

Please take look at the change and let me know your feedback.

harendra-kumar commented:
@NorfairKing, @sergv can you please take a look.

harendra-kumar commented on CHANGELOG:
```diff
@@ -1,6 +1,21 @@
+		You may have to replace invalid forms of extensions with valid ones.
+		Especially:
+		* "foo" was earlier equivalent to ".foo" but is now invalid
+		* ".foo.bar" is now invalid, instead set ".foo" and ".bar" separately
```
Think about this law:

```
(fileExtension . fromJust . addFileExtension ext) file == ext
```
If we allow `.tar.gz` to be set then `fileExtension` will return `.gz` and this law will not hold. However, if we set one extension at a time then this law always holds.  But it is not strictly necessary that we obey this law, we have the option to allow extensions like`.tar.gz` and say that this is not a law.

NorfairKing replied:
Sounds good to me.

harendra-kumar replied:
I am trying to interpret your response. Does that mean we should allow `.tar.gz`? In fact I am tending towards allowing it because majority of users will intuitively think that it should be allowed. And this law does not sound very important to me.

NorfairKing commented:
I would prefer to see property tests for this kind of change before we go on with it.

harendra-kumar commented on src/Path/Include.hs:
```diff
@@ -334,47 +346,105 @@ dirname (Path "") = Path ""
 --
+-- >>> fileExtension $(mkRelFile "name.foo"     ) == ".foo"
+-- >>> fileExtension $(mkRelFile "name.foo."    ) == ".foo."
+-- >>> fileExtension $(mkRelFile "name.foo.."   ) == ".foo.."
```
Dots in trailing position of an extension are valid, they can be semantically allowed without violating any rule. In fact they are necessary so that we can consistently and unambiguously compose/decompose extensions of this form:

```
name.foo...bar
```

The extension in this case is a composition of `.foo..` + `.bar`, there is no ambiguity. Notice that, if we drop the extension from `name.foo...bar` we will get `name.foo..` and it has an extension `.foo..`, everything is consistent.

harendra-kumar replied:
I allowed trailing dots in a separate commit, maybe you were looking at the first commit only where `.foo..` was invalid?

harendra-kumar commented on src/Path/Include.hs:
```diff
@@ -334,47 +346,105 @@ dirname (Path "") = Path ""
+-- >>> fileExtension $(mkRelFile "name."        ) == ""
+-- >>> fileExtension $(mkRelFile "name.."       ) == ""
+-- >>> fileExtension $(mkRelFile ".name"        ) == ""
+-- >>> fileExtension $(mkRelFile "..name"       ) == ""
```
We have already discussed and acknowledged in the other thread that `.name` and `..name` are filenames. So I guess you are only worried about `name.` and `name..` here, both of which are actually specific instances of the same general case. 

Well, we have to acknowledge the existence of extensions either way. For example you will either have to say that `name.` is a file with an extension `.` or that `name.` is just a file without an extension. You cannot have it both ways or leave it to be interpreted whatever way. Choosing `name.` to be a filename plays nicely with the laws, keeps everything consistent and unambiguous.

harendra-kumar commented:
> I would prefer to see property tests for this kind of change before we go on with it.

Can you elaborate a bit more on the nature of the tests or give a few examples, where to add and testing what specifically? BTW, I have added tests creating a large number of permutations and combinations resulting in more than 400 tests.

sergv commented on src/Path/Include.hs:
```diff
@@ -334,47 +346,105 @@ dirname (Path "") = Path ""
+
+        -- All "."s is not a valid extension
+        let ys = dropWhile FilePath.isExtSeparator (reverse xs)
+        when (ys == []) $
```
Same observation regarding emptiness checking as for `xs`.

harendra-kumar commented on src/Path/Include.hs:
```diff
@@ -334,47 +346,105 @@ dirname (Path "") = Path ""
 
+-- | Split the given file path into a filename and an extension. If the
+-- filename has no extension it returns "" as extension.
+splitExtension :: Path b File -> (Path b File, String)
```
Yes, I totally agree. However I wanted this to be done as a separate change once we merge this one to keep this discussion/changes in this PR as short as possible. This PR does not change or introduce any new APIs at all, just changes the semantics of existing extension APIs.

harendra-kumar commented on src/Path/Include.hs:
```diff
@@ -334,47 +346,105 @@ dirname (Path "") = Path ""
+            throwM $ InvalidExtension ex
+
+        -- just a "." is not a valid extension
+        when (xs == []) $
```
Its just a matter of taste I guess. Here we are just pattern matching on a constructor so the code is as efficient and simple as it can be. But yes, `null` can be used if we want to follow a general convention across containers.

harendra-kumar commented:
@NorfairKing any luck with this one? Let me know if there is anything that needs to be done to get this merged. I am afraid I might just lose context and it might just get dropped out of my radar. I think this is a good change to have. This provides you a way to have an `splitExtension` functionality.

harendra-kumar commented:
> Can you elaborate a bit more on the nature of the tests or give a few examples, where to add and 
> testing what specifically?

As I asked before can you help me out a little bit here?

NorfairKing commented:
For example, if your law is ` (fileExtension . fromJust . addFileExtension ext) f == ext`, then you should have this property test:
``` haskell
forAllValid $ \ext -> 
  (fileExtension . fromJust . addFileExtension ext) f `shouldBe` ext
```

harendra-kumar commented:
It may take me some time to get back to it again. Please feel free to add the tests yourself if you feel like doing so.

harendra-kumar commented:
Fixed the redundant `shouldBe`.

NorfairKing commented on test/ValidityTest.hs:
```diff
@@ -152,7 +152,7 @@ extensionsSpec = do
              forAllShrink genValid shrinkValidExtension $ \(Extension ext) ->
                 (toFilePath . fromJust . addFileExtension ext) file
-                 `shouldBe` toFilePath file ++ ext
+                 == toFilePath file ++ ext
```
`shouldBe` was _not_ redundant here.
Now your error messages will be unhelpful.

harendra-kumar commented:
@NorfairKing one problem that I see with this change is that the breaking change will cause runtime failures and not compile time issues. So this is dangerous in that respect. We can allow `.tar.gz` but any instances of setting `tar` instead of `.tar` will still cause failures at runtime.  I am okay to drop it if this is considered unacceptable.

One way to manage it could be to introduce new APIs and deprecate the old ones. That way we can avoid runtime failures. We will have to have new versions of `addFileExtension` and `setFileExtension`. We can change those to `addExtension` and `setExtension` (or `replaceExtension`) or something like that.

NorfairKing commented:
> @NorfairKing one problem that I see with this change is that the breaking change will cause runtime failures and not compile time issues. 

That's to be expected for major-version bumps. I'm fine with it.

harendra-kumar commented:
@chrisdone are you okay with a runtime breaking change, see the changelog for what breaks? The alternative is deprecation and renaming of two functions as suggested above.

mrkkrp commented:
> The alternative is deprecation and renaming of two functions as suggested above

I'd like to vote for this solution, because it's more explicit and with this approach it's harder for end users to get in trouble.

harendra-kumar commented:
I agree, sudden change of runtime behavior may be surprising and may evoke WTFs. Not sure how many users are exposed to this. Deprecation is easier than inflicting that pain on users.

harendra-kumar commented:
So what should be the names of the new APIs, I propose the following:

* `addFileExtension` is deprecated and replaced by `addExtension`
* `setFileExtension` is deprecated and replaced by `replaceExtension`. I am explicitly adding the name `replace` in this API because I always get confused whether this is going to append or replace. In case there is no extension it will still add an extension but that is easier to resolve compared to the confusion with the previous name. @mrkkrp what do you think?

mrkkrp commented:
I don't have a strong opinion regarding the names, yours looks fine.

harendra-kumar commented:
There are two operators as well `<.>` and `-<.>` corresponding to `addFileExtension` and `setFileExtension`. What do we do with those? One possibility is to deprecate the operators now and revive them later with the new behavior.

NorfairKing commented:
The deprecation option sounds good to me.

> One possibility is to deprecate the operators now and revive them later with the new behavior.

I'm all for removing the operators entirely.

harendra-kumar commented:
@NorfairKing @mrkkrp I guess this can be merged now. 

Once this is merged I have another small change that exposes the `splitExtension` operation, it will complete the extension operations' repertoire.
